### PR TITLE
cpp object detection demo: add wrapper for retinaface models

### PR DESCRIPTION
**Pull request by ivanvikhrev**
C++ wrapper for retinaface-anti-cov and retinaface-resnet50 models for object detection demo.

### CODE REVIEW

**Comment by eaidova** [comment]
Jenkins ok to test

**Review by vladimir-dudnik — APPROVED** [review]

File: demos/common/models/include/models/detection_model_retinaface.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    static const int INIT_VECTOR_SIZE = 200;
+    /// Loads model and performs required initialization
+    /// @param model_name name of model to load
+    /// @param confidenceThreshold - threshold to eleminate low-confidence detections.
+    /// Any detected object with confidence lower than this threshold will be ignored.
+    /// @param useAutoResize - if true, image will be resized by IE.
+    /// @param labels - array of labels for every class. If this array is empty or contains less elements
+    /// than actual classes number, default "Label #N" will be shown for missing items.
+    ModelRetinaFace(const std::string& model_name, float confidenceThreshold, bool useAutoResize, float boxIOUThreshold);
+    std::unique_ptr<ResultBase> postprocess(InferenceResult& infResult);
```
Comment (Wovchena) [inline demos/common/models/include/models/detection_model_retinaface.h:46]:
```suggestion
    std::unique_ptr<ResultBase> postprocess(InferenceResult& infResult) override;
```

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+ModelRetinaFace::ModelRetinaFace(const std::string& modelFileName, float confidenceThreshold, bool useAutoResize, float boxIOUThreshold)
+    : DetectionModel(modelFileName, confidenceThreshold, useAutoResize, {"Face"}),  // Default label is "Face"
+    boxIOUThreshold(boxIOUThreshold), maskThreshold(0.8f), shouldDetectMasks(false), landmarkStd(1.0f),
+    anchorCfg({ {32, { 32, 16 }, 16, { 1 }},
+              { 16, { 8, 4 }, 16, { 1 }},
+              { 8, { 2, 1 }, 16, { 1 }} }) {
+    generateAnchorsFpn();
+}
+
+void ModelRetinaFace::prepareInputsOutputs(InferenceEngine::CNNNetwork& cnnNetwork) {
```
Comment (Wovchena) [inline demos/common/models/src/detection_model_retinaface.cpp:33]: You already have `using namespace InferenceEngine`. Remove `InferenceEngine::` or `using namespace InferenceEngine;`.

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    for (auto ra : ratioAnchors) {
+        auto addon = scaleEnum(ra, scales);
+        retVal.insert(retVal.end(), addon.begin(), addon.end());
+    }
+    return retVal;
+}
+
+void ModelRetinaFace::generateAnchorsFpn() {
+    auto cfg = anchorCfg;
+    std::sort(cfg.begin(), cfg.end(), [](AnchorCfgLine& x, AnchorCfgLine& y) {return x.stride > y.stride; });
```
Comment (Wovchena) [inline demos/common/models/src/detection_model_retinaface.cpp:156]:
```suggestion
    std::sort(cfg.begin(), cfg.end(), [](const AnchorCfgLine& x, const AnchorCfgLine& y) { return x.stride > y.stride; });
```

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+    std::vector<ModelRetinaFace::Anchor> anchors;
+    for (auto cfgLine : cfg) {
+        auto anchors = generateAnchors(cfgLine.baseSize, cfgLine.ratios, cfgLine.scales);
+        anchorsFpn.emplace(cfgLine.stride,anchors);
```
Comment (Wovchena) [inline demos/common/models/src/detection_model_retinaface.cpp:161]:
```suggestion
    for (auto cfgLine : cfg) {
        anchorsFpn.emplace(cfgLine.stride, generateAnchors(cfgLine.baseSize, cfgLine.ratios, cfgLine.scales));
```

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+        const auto bboxRaw = infResult.outputsData[separateOutputsNames[OT_BBOX][idx]];
+        const auto scoresRaw = infResult.outputsData[separateOutputsNames[OT_SCORES][idx]];
+        const auto landmarksRaw = infResult.outputsData[separateOutputsNames[OT_LANDMARK][idx]];
+        auto sz = bboxRaw->getTensorDesc().getDims();
+        auto height = sz[2];
+        auto width = sz[3];
+        auto s = anchorCfg[idx].stride;
+        auto anchorNum = anchorsFpn[s].size();
+
+        //--- Creating strided anchors plane at first start
```
Comment (Wovchena) [inline demos/common/models/src/detection_model_retinaface.cpp:328]: Shouldn't this be in `prepareInputsOutputs()`? It looks it depends only on output shape of OT_BBOX.

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: demos/common/models/include/models/detection_model_retinaface.h [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    };
+
+    std::vector <std::string> separateOutputsNames[OT_MAX];
+    const std::vector<AnchorCfgLine> anchorCfg;
+    std::map<int, std::vector <Anchor>> anchorsFpn;
+    std::vector<std::vector<Anchor>> anchors;
+
+    void generateAnchorsFpn();
+    virtual void prepareInputsOutputs(InferenceEngine::CNNNetwork& cnnNetwork) override;
+
```
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/common/models/include/models/detection_model_retinaface.h]: ok

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+
+    InferenceEngine::OutputsDataMap outputInfo(cnnNetwork.getOutputsInfo());
+
+    std::vector<int> outputsSizes[OT_MAX];
+    for (auto& output : outputInfo) {
+        output.second->setPrecision(InferenceEngine::Precision::FP32);
+        output.second->setLayout(InferenceEngine::Layout::NCHW);
+        outputsNames.push_back(output.first);
+
+        EOutputType type= OT_MAX;
```
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: ok

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+        auto predH = exp(dh) * anchors[i].getHeight();
+
+        bboxes->push_back({ static_cast<float>(predCtrX - 0.5f * (predW - 1.0f)), static_cast<float>(predCtrY - 0.5f * (predH - 1.0f)),
+           static_cast<float>(predCtrX + 0.5f * (predW - 1.0f)), static_cast<float>(predCtrY + 0.5f * (predH - 1.0f)) });
+    }
+}
+
+
+void filterLandmarks(std::vector<cv::Point2f>* landmarks, const std::vector<size_t>& indices, InferenceEngine::MemoryBlob::Ptr rawData,
+    int anchorNum, const std::vector<ModelRetinaFace::Anchor>& anchors, const float landmarkStd) {
```
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: ok

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+
+        if (shouldDetectMasks) {
+            const auto masksRaw = infResult.outputsData[separateOutputsNames[OT_MASKSCORES][idx]];
+            filterMasksScores(&masks, validIndices, masksRaw, anchorNum);
+        }
+    }
+// --------------------------- Apply Non-maximum Suppression ----------------------------------------------------------
+    auto keep = nms(bboxes, scores, boxIOUThreshold);
+
+// --------------------------- Create detection result objects --------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: ok

File: demos/common/models/include/models/detection_model_retinaface.h (hunk 1 above)
Comment (ivanvikhrev) [reply demos/common/models/include/models/detection_model_retinaface.h]: ok

File: demos/common/models/src/detection_model_retinaface.cpp (hunk 3 above)
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: ok

File: demos/common/models/src/detection_model_retinaface.cpp (hunk 4 above)
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: ok

File: demos/object_detection_demo/main.cpp [hunk 10]
```diff
@@ -181,6 +182,13 @@ cv::Mat renderDetectionData(const DetectionResult& result) {
         cv::rectangle(outputImg, obj, cv::Scalar(0, 0, 255));
     }
 
+    try {
+        for (auto lmark : result.asRef<RetinaFaceDetectionResult>().landmarks) {
+            cv::circle(outputImg, lmark, 2, cv::Scalar(0, 255, 255), -1);
+        }
+    }
+    catch (...) {}
```
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/object_detection_demo/main.cpp:190]: `result.asRef<RetinaFaceDetectionResult>().landmarks `will cause error` Bad dynamic_cast!` for other models.

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+
+    std::vector<ModelRetinaFace::Anchor> anchors;
+    for (auto cfgLine : cfg) {
+        auto anchors = generateAnchors(cfgLine.baseSize, cfgLine.ratios, cfgLine.scales);
+        anchorsFpn.emplace(cfgLine.stride,anchors);
+    }
+}
+
+
+std::vector<size_t> thresholding(InferenceEngine::MemoryBlob::Ptr rawData, const int anchorNum, const float confidenceThreshold) {
```
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: Done.

File: demos/common/models/src/detection_model_retinaface.cpp [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+                    indices.push_back((y * sz[3] + z) * restAnchors + (x - anchorNum));
+                }
+            }
+        }
+    }
+
+    return indices;
+}
+
+void filterScores(std::vector<float>* scores, const std::vector<size_t>& indices, InferenceEngine::MemoryBlob::Ptr rawData, const int anchorNum) {
```
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: Done.

File: demos/object_detection_demo/main.cpp [hunk 13]
```diff
@@ -38,8 +38,9 @@
 #include <pipelines/config_factory.h>
 #include <pipelines/metadata.h>
 #include <models/detection_model_faceboxes.h>
-#include <models/detection_model_yolo.h>
+#include <models/detection_model_retinaface.h>
 #include <models/detection_model_ssd.h>
+#include <models/detection_model_yolo.h>
 
 static const char help_message[] = "Print a usage message.";
 static const char at_message[] = "Required. Architecture type: ssd or yolo";
```
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/object_detection_demo/main.cpp]: Done.

File: demos/common/models/src/detection_model_retinaface.cpp (hunk 2 above)
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: Removed using namespace.

File: demos/common/models/src/detection_model_retinaface.cpp (hunk 5 above)
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: Done.

File: demos/common/models/src/detection_model_retinaface.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/common/models/src/detection_model_retinaface.cpp]: `filterMasksScores()` should be updated as well.

File: demos/object_detection_demo/main.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/object_detection_demo/main.cpp]: The help message in README should be updated

File: demos/object_detection_demo/main.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/object_detection_demo/main.cpp]: Catch `const std::bad_cast&` then.

File: demos/common/cpp/models/include/models/detection_model_retinaface.h [hunk 14]
```diff
@@ -0,0 +1,76 @@
+/*
+// Copyright (C) 2018-2020 Intel Corporation
```
Comment (Wovchena) [inline demos/common/cpp/models/include/models/detection_model_retinaface.h:2]:
```suggestion
// Copyright (C) 2020-2021 Intel Corporation
```

File: demos/common/cpp/models/include/models/detection_model_retinaface.h [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        OT_MAX
+    };
+
+    std::vector <std::string> separateOutputsNames[OT_MAX];
+    const std::vector<AnchorCfgLine> anchorCfg;
+    std::map<int, std::vector <Anchor>> anchorsFpn;
+    std::vector<std::vector<Anchor>> anchors;
+
+    void generateAnchorsFpn();
+    virtual void prepareInputsOutputs(InferenceEngine::CNNNetwork& cnnNetwork) override;
```
Comment (Wovchena) [inline demos/common/cpp/models/include/models/detection_model_retinaface.h:75]:
```suggestion
    void prepareInputsOutputs(InferenceEngine::CNNNetwork& cnnNetwork) override;
```

File: demos/common/cpp/models/src/detection_model_retinaface.cpp [hunk 16]
```diff
@@ -0,0 +1,389 @@
+/*
+// Copyright (C) 2018-2020 Intel Corporation
```
Comment (Wovchena) [inline demos/common/cpp/models/src/detection_model_retinaface.cpp:2]:
```suggestion
// Copyright (C) 2020-2021 Intel Corporation
```
These are nearly added files

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: demos/common/cpp/models/src/detection_model_retinaface.cpp (hunk 16 above)
Comment (ivanvikhrev) [reply demos/common/cpp/models/src/detection_model_retinaface.cpp:2]: ok

File: demos/common/cpp/models/include/models/detection_model_retinaface.h (hunk 15 above)
Comment (ivanvikhrev) [reply demos/common/cpp/models/include/models/detection_model_retinaface.h:75]: ok

File: demos/common/cpp/models/include/models/detection_model_retinaface.h (hunk 14 above)
Comment (ivanvikhrev) [reply demos/common/cpp/models/include/models/detection_model_retinaface.h:2]: ok

File: demos/common/models/src/detection_model_retinaface.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/common/models/src/detection_model_retinaface.cpp]: Ok, missed it first time.

File: demos/object_detection_demo/main.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/object_detection_demo/main.cpp]: Updated

File: demos/object_detection_demo/main.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/object_detection_demo/main.cpp]: Changed

File: demos/object_detection_demo/main.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/object_detection_demo/main.cpp]: You don't need coma before "or".

[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/object_detection_demo/main.cpp]: I'll change it with another PR wtih fixes for faceboxes wrapper soon.

[the remark this answers is not in the document]
Comment (ivanvikhrev) [reply demos/object_detection_demo/main.cpp]: Well, seems it has to be done here due to merge conflict.

**Review by Wovchena — APPROVED** [review]

File: demos/common/cpp/models/src/detection_model_retinaface.cpp [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+    }
+    return retVal;
+}
+
+std::vector<ModelRetinaFace::Anchor> generateAnchors(const int baseSize, const std::vector<int>& ratios, const std::vector<int>& scales) {
+    ModelRetinaFace::Anchor baseAnchor{ 0.0f, 0.0f, baseSize - 1.0f, baseSize - 1.0f };
+    auto ratioAnchors = ratioEnum(baseAnchor, ratios);
+    std::vector<ModelRetinaFace::Anchor> retVal;
+
+    for (auto ra : ratioAnchors) {
```
Comment (IRDonch) [inline demos/common/cpp/models/src/detection_model_retinaface.cpp:167]:
```suggestion
    for (const auto& ra : ratioAnchors) {
```

File: demos/common/cpp/models/src/detection_model_retinaface.cpp [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+        retVal.insert(retVal.end(), addon.begin(), addon.end());
+    }
+    return retVal;
+}
+
+void ModelRetinaFace::generateAnchorsFpn() {
+    auto cfg = anchorCfg;
+    std::sort(cfg.begin(), cfg.end(), [](const AnchorCfgLine& x, const AnchorCfgLine& y) { return x.stride > y.stride; });
+
+    for (auto cfgLine : cfg) {
```
Comment (IRDonch) [inline demos/common/cpp/models/src/detection_model_retinaface.cpp:178]:
```suggestion
    for (const auto& cfgLine : cfg) {
```

**Review by IRDonch — CHANGES_REQUESTED** [review]

**Review by IRDonch — APPROVED** [review]

**Pull request merged**